(feat:no-unnecessary-act): adds no-unnecessary-act rule, closes #259

alessbell opened the pull request:
I was starting to spin my wheels a bit so I wanted to open the PR to get some feedback 🙂 Closes #259. 

I'm having trouble detecting all three of these cases:
- empty functions
- arrow functions with block bodies
- arrow functions with concise bodies

My interpretation of the business logic is:
Find all instances of `act` (regardless of how `act` is imported, `ReactTestUtils.act` should be treated the same as `act`) and check whether the function passed to `act`:
  1. has an empty body = act unnecessary
  2. contains a non-Testing Library function call = act is permitted

Five test cases are still failing. Will circle back around to this later today but wanted to open it up for any and all advice, thanks @Belco90 !

Belco90 commented:
Nice! I'll take a look during the weekend.

It would be nice if you can update your branch again with `v4` branch since we already merged the fix for running the pipeline correctly.

MichaelDeBoey commented on docs/rules/no-unnecessary-act.md:
```diff
@@ -0,0 +1,26 @@
+
+## Rule Details
+
+This rule disallows the usage of `act` when using it to wrap Testing Library functions, or functions with an empty body, which suppresses valid warnings. For more information, see [https://kcd.im/react-act](https://kcd.im/react-act).
```
```suggestion
This rule disallows the usage of `act` when using it to wrap Testing Library
functions, or functions with an empty body, which suppresses valid warnings. For
more information, see https://kcd.im/react-act.
```

MichaelDeBoey commented on tests/lib/rules/no-unnecessary-act.test.ts:
```diff
@@ -0,0 +1,185 @@
+      `
+    },
+    {
+      code: `
```
Should be invalid, no? 🤔

Belco90 replied:
Confirmed: this is valid.

MichaelDeBoey commented on tests/lib/rules/no-unnecessary-act.test.ts:
```diff
@@ -0,0 +1,185 @@
+      `
+    },
+    {
+      code: `
```
Should be invalid, no? 🤔

Belco90 replied:
Confirmed: this is valid.

Belco90 commented on docs/rules/no-unnecessary-act.md:
````diff
@@ -0,0 +1,26 @@
+Examples of **incorrect** code for this rule:
+
+```js
+import { fireEvent, act } from '@testing-library/react';
````
I would actually use both examples: with `fireEvent` to clarify it's a Testing Library util already wrapped with `act`, and `userEvent` to clarify the same since it's imported from a different package.

Belco90 reviewed (changes requested):
Thanks for your PR @alessbell ! I'm flagging it as WIP until you finish it.

It looks good so far! However, I've left some comments about how to detect renamed `act` properly. You may need to change the implementation of the `CallExpression` listener a bit to get this properly.

Regarding the detection of empty functions, you can get some inspiration from `no-wait-for-empty-callback` where we actually report empty functions :)

I don't know if it could be interesting to report the `act` here in two different ways:
- for `act` with empty functions: look for a `CallExpression`, then check it's an `act` to be reported, and finally check if its callback is empty
- for `act` containing non Testing Library calls: look for any `CallExpression, then check it's not coming from Testing Library (you have a helper available for it), and if you find one then flag the closest parent `act` called as valid. I'm not sure if this approach is easier or not than the opposite one (look for `CallExpression` that are `act` to be reported and then check for calls inside its callback) you are following to be honest, it's just an idea in case it helps. Remember [eslint-utils](https://eslint-utils.mysticatea.dev/api/ast-utils.html) is part of `@typescript-eslint/experimental-utils` we use, in case there is any function there it helps you to detect something.

Apart from the comments I've added throughout the PR, I think you should add more tests (specially without aggressive reporting, you can see the `no-debug` PR that I mentioned in one comment) and some extra examples in the rule doc.

Belco90 commented on tests/lib/rules/no-unnecessary-act.test.ts:
```diff
@@ -0,0 +1,185 @@
+      `,
+      errors: [
+        {
+          messageId: 'noUnnecessaryAct'
```
I forgot to mention: can you check line and column for reported errors?

Belco90 commented on tests/lib/rules/no-unnecessary-act.test.ts:
```diff
@@ -0,0 +1,185 @@
+
+const ruleTester = createRuleTester({
+  ecmaFeatures: {
+    jsx: true
```
Our `createRuleTester` already enables `jsx` so you don't need to set it here :)

gndelia commented on lib/rules/no-unnecessary-act.ts:
```diff
@@ -0,0 +1,85 @@
+      let isEmpty = false;
+      let callsTL = false;
+      const callsOnlyTLFns = (body: Array<TSESTree.Node>): boolean =>
+        body.every((node: TSESTree.ExpressionStatement) => {
+          if (
+            isCallExpression(node.expression) &&
+            isMemberExpression(node.expression.callee) &&
+            isCallExpression(node.expression.callee.object) &&
+            helpers.isNodeComingFromTestingLibrary(node.expression.callee)
+          ) {
+            return true;
+          }
+          return false;
+        });
```
really minor suggestion, but you can drop the if clause if you want and just return the expression
```suggestion
        body.every((node: TSESTree.ExpressionStatement) => 
            isCallExpression(node.expression) &&
            isMemberExpression(node.expression.callee) &&
            isCallExpression(node.expression.callee.object) &&
            helpers.isNodeComingFromTestingLibrary(node.expression.callee)
        );
```

Belco90 commented:
@alessbell I just realized this PR has been closed because it pointed to `v4` branch which has been closed in favor of `beta` branch. I can't reopen it myself, but if you can edit the base branch to be `beta` it should be fine. Also, you can create another PR too. Sorry about it, I didn't know the branch change would cause this :(

We have already a beta released, so I have more spare time now. Can I help with something here? After having to migrate 20+ rules from the plugin, I really feel into writing ESLint rules so ping me if you want me to give you a hand here!

alessbell commented:
@Belco90 thanks for your patience, and offer to help! It's not letting me edit the base branch for some reason, so I think I'll open a new PR this weekend and will let you know if I have any questions once I've addressed the comments left here 👍
